README.txt
See the docs/index.txt file for documentation links.



CHANGELOG.txt
Changes
=======

0.4.5

* Bumped SQLAlchemy version to 0.5
* Fixed ticket 67 to add Python 2.4 support, patch by Kumar McMillan

**NOTE**: The old trac system is no longer used. Any outstanding bugs should be
submitted **with patches and tests** to the AuthKit mailing list.

0.4.4

* Added patch by Phil Kershaw for Attribute exhange in OpenID
* Added patch by Phil Kershaw for hiding the continue button during
  OpenID redirection
* Changed cookie handling so that if nouserincookie is set, the user data as
  well as the username is stored in the session rather than the cookie
* Added a user_data function to the form method so you can set arbitrary
  user data messages via the same mechanism as other options

0.4.3

* Updated authkit.authenticate.form to handle the suggest change in #61
* Added powerpack adaptors
* Changed the multi middleware to return the WSGI response rather than
  iterating over it and yielding results.
* Fixed #50 OpenID problem and tested on blogger.com
* Added the ability to pass the environ dictionary to render() functions
* Added bad cookie customisation options to fix #65 but also to allow
  bad cookie template customisation. See the docstring of the 
  authkit.authenticate.cookie module.
* Added a new algorithm based on ideas from #61 to guess the correct action
  for the form produced by the form middleware but also added support for
  an authkit.form.action option which allows you to manually override 
  AuthKit's guess. (The OpenID middleware calls this baseurl)
* Added user_set_password() methods to users API. Fixes #64.
* Removed arabic letters from the form handler. Fixes #40.
* Added headers to the form handling for IE. Fixes #54.
* Adding SQLAlchemy 0.5 support, uses session.add() instead of session.save()
* The multi handler now handles WSGI applications implemented as iterators,
  it already supported generators. The multi2.py example demonstrates this.
* Set the pylons.status_code_redirect environment variable on all redirected 
  AuthKit responses
* Set the pylons.error_call environment variable on all redirected AuthKit
  responses
* Trigger an error when trying to use server-side password encryption with 
  digest authentication

0.4.2

* Applied most of #59 and tested the two OpenID examples
* Added support for WebOb HTTPExceptions to remove a deprecation warning in 
  Pylons 0.9.7
* Fixed a missing md5 import in the cookie module
* Applied patch in ticket #52
   
0.4.1 

* Added ``setup.fakeuser`` option which automatically sets the REMOTE_USER
  so that it appears someone has signed in. Useful with setup.enable = False

* The ``setup.enable = false`` option now also disables  authorisation          
  checks (reported by Rick Flosi)                                               

* Applied patch from Pawel Niewiadomski to fix #53
* Changed the import of the openid.sreg module to openid.extensions.sreg
* Fixed the encoding of the form.py file
* Updated the examples to use the latest syntax
* Updated the tests for the new SQLAlchemy drivers
* Added Daniel Pronych's SQLAlchemy drivers but with significant changes
* Fixed a bug due to a change in the latest version of Python OpenID
  so that AuthKit OpenID works with Yahoo sign-ins. Phil Kershaw #50
* Updated the user/database example, it now works #43
* Updated user tokens code to fix #17
* Updated authkit.authenticate.multi so that it should fix #41 and pass
  the WSGI writable from start_response() correctly.
* Added a form.method option so you can choose GET authentication for
  cases when another piece of middleware has already parsed the form 
  variables (eg with repoze?).
* Fixed bug in form action generation for non-standard ports
* Fixed bug reported by Sam Gentle where remote addr is obtained from 
  X_FORWARDED_FOR after multiple proxies.
* AuthKit form authentication now picks up HTTP_X_FORWARDED_HOST and
  HTTP_X_FORWARDED_PORT when generating an action. This allows you to run
  an AuthKit app on port 80, proxied from 443 as long as you set up these
  two (slightly unstandard) variables.
* Fixed #38, cookie sign out path should match the path specified in 
  the config file.
* Fixed #37, missing import of sys in digest authentication
* Updated SQLAlchemy code to use SQLAlchemyManager. Needs installing 
  manually with ``easy_install SQLAlchemyManager``.
* Added a user management api_version attribute and changed the API so that
  the users object is set up on each request and recieves an environ
  argument.
* OpenID middleware now no longer sets up beaker middleware itself. This 
  should now be done manually in the middleware stack. See the example in 
  examples/docs/open_id.py
* OpenID support now upgraded to use 2.0 (from Dalius Dobravolskas)
* Fixed encrypt typo with postgres users driver
* The cookie middleware now has a nouserincookie option which forces the 
  middleware to store the username in a Beaker session rather than in plain
  text in the cookie. See examples/docs/form_no_user_in_cookie.py for an
  example of its use.
* Form authentication method now defaults to 200 OK rather than 401 when 
  the sign in form is displayed so that it works with Safari 3 Beta. 
* The cookie middleware Bad Ticket page now also uses 200 OK, also to 
  support Safari 3 Beta.
* The cookie middleware bad ticket now logs to debug rather than error to 
  avoid the ``No handlers could be found for logger 
  "authkit.authenticate.cookie"`` message you get with a bad cookie if no
  error logging is specifically set up.
* Added a user management api_version attribute and changed the API so that
  the users object is set up on each request and recieves an environ
  argument.
* Fixed encrypt typo with postgres users driver
* Renamed the config_paste option to app_conf in authenticate middleware. If 
  you get an "No authkit.setup.method was specified" error when you are sure
  it is specified, this might be why.

0.4.0 

* Added support for encrypted passwords
* Fixed the IE7 bug in digest middleware

authkit/authenticate/form.py
# -*- coding: utf-8 -*-

"""Form and cookie based authentication middleware

As with all the other AuthKit middleware, this middleware is described in
detail in the AuthKit manual and should be used via the
``authkit.authenticate.middleware`` function.

The option form.status can be set to "200 OK" if the Pylons error document
middleware is intercepting the 401 response and just showing the standard 401
error document. This will not happen in recent versions of Pylons (0.9.6)
because the multi middleware sets the environ['pylons.error_call'] key so that
the error documents middleware doesn't intercept the response.

From AuthKit 0.4.1 using 200 OK when the form is shown is now the default. 
This is so that Safari 3 Beta displays the page rather than trying to 
handle the response itself as a basic or digest authentication.
"""

from paste.auth.form import AuthFormHandler
from paste.request import parse_formvars
from authkit.authenticate import get_template, valid_password, \
   get_authenticate_function, strip_base, RequireEnvironKey, \
   AuthKitAuthHandler
from authkit.authenticate.multi import MultiHandler, status_checker

import inspect
import logging
import urllib
log = logging.getLogger('authkit.authenticate.form')

def user_data(state):
    return 'User data string'

def template(method=False):
    t = """\
<html>
  <head><title>Please Sign In</title></head>
  <body>
    <h1>Please Sign In</h1>
    <form action="%s" method="post">
      <dl>
        <dt>Username:</dt>
        <dd><input type="text" name="username"></dd>
        <dt>Password:</dt>
        <dd><input type="password" name="password"></dd>
      </dl>
      <input type="submit" name="authform" value="Sign In" />
    </form>
  </body>
</html>
"""
    if method is not False:
        t = t.replace('post', method)
    return t

class AttributeDict(dict):
    def __getattr__(self, name):
        if not self.has_key(name):
            raise AttributeError('No such attribute %r'%name)
        return self.__getitem__(name)

    def __setattr__(self, name, value):
        raise NotImplementedError(
            'You cannot set attributes of this object directly'
        )

class FormAuthHandler(AuthKitAuthHandler, AuthFormHandler):
    def __init__(
        self, 
        app,
        charset=None,
        status="200 OK",
        method='post',
        action=None,
        user_data=None,
        **p
    ):
        AuthFormHandler.__init__(self, app, **p)
        self.status = status
        self.content_type = 'text/html'
        self.charset = charset
        if self.charset is not None:
            self.content_type = self.content_type + '; charset='+charset
        self.method = method
        self.action = action
        self.user_data = user_data
    
    def on_authorized(self, environ, start_response):
        if self.user_data is not None:
            state = environ.get('wsgiorg.state')
            if not state:
                environ['wsgiorg.state'] = state = AttributeDict()
                state['environ'] = environ
                state['start_response'] = start_response
            environ['paste.auth_tkt.set_user'](userid=environ['REMOTE_USER'], user_data=self.user_data(state))
        else:
            environ['paste.auth_tkt.set_user'](userid=environ['REMOTE_USER'])
        return self.application(environ, start_response)
        
    def __call__(self, environ, start_response):
        # Shouldn't ever allow a response if this is called via the 
        # multi handler
        username = environ.get('REMOTE_USER','')
        formvars = parse_formvars(environ, include_get_vars=True)
        username = formvars.get('username')
        password = formvars.get('password')
        if username and password:
            if self.authfunc(environ, username, password):
                log.debug("Username and password authenticated successfully")
                environ['AUTH_TYPE'] = 'form'
                environ['REMOTE_USER'] = username
                environ['REQUEST_METHOD'] = 'GET'
                environ['CONTENT_LENGTH'] = ''
                environ['CONTENT_TYPE'] = ''
                del environ['paste.parsed_formvars']
                return self.on_authorized(environ, start_response)
            else:
                log.debug("Username and password authentication failed")
        else:
            log.debug("Either username or password missing")
        action = self.action or construct_url(environ)
        log.debug("Form action is: %s", action)

        # Inspect the function to see if we can pass it anything useful:
        args = {}
        kargs = {'environ':environ}
        if environ.has_key('gi.state'):
            kargs['state'] = environ['gi.state']
        for name in inspect.getargspec(self.template)[0]:
            if kargs.has_key(name):
                args[name] = kargs[name]
        if self.method != 'post':
            args['method'] = self.method
        content = self.template(**args) % (action)
        if self.charset is not None:
            content = content.encode(self.charset)
            
        writable = start_response(
            self.status,
            [
                ('Content-Type', self.content_type),
                ('Content-Length', str(len(content))),
                # Added for IE compatibility - see #54
                ('Pragma', 'no-cache'),
                ('Cache-Control', 'no-cache'),
            ]
        )
        return [content]

def construct_url(environ, with_query_string=True, with_path_info=True,
                  script_name=None, path_info=None, querystring=None):
    """Reconstructs the URL from the WSGI environment.

    You may override SCRIPT_NAME, PATH_INFO, and QUERYSTRING with
    the keyword arguments.

    """
    url = '://'
    host = environ.get('HTTP_X_FORWARDED_HOST', environ.get('HTTP_HOST'))
    port = None
    if ':' in host:
        host, port = host.split(':', 1)
    else:
        # See if the request is proxied
        host = environ.get('HTTP_X_FORWARDED_HOST', environ.get('HTTP_X_FORWARDED_FOR'))
        if host is not None:
            # Request was proxied, get the correct data
            host = environ.get('HTTP_X_FORWARDED_HOST')
            port = environ.get('HTTP_X_FORWARDED_PORT')
            if port is None and environ.get('HTTP_X_FORWARDED_SSL') == 'on':
                port = '443'
            if not port:
                log.warning(
                    'No HTTP_X_FORWARDED_PORT or HTTP_X_FORWARDED_SSL found '
                    'in environment, cannot '
                    'determine the correct port for the form action. '
                ) 
            if not host:
                log.warning(
                    'No HTTP_X_FORWARDED_HOST found in environment, cannot '
                    'determine the correct hostname for the form action. ' 
                    'Using the value of HTTP_HOST instead.'
                )   
                host = environ.get('HTTP_HOST')
        else:
            # Request was not proxied
            if environ['wsgi.url_scheme'] == 'https':
                port = 443
            if host is None:
                host = environ.get('HTTP_HOST')
            if port is None:
                port = environ.get('SERVER_PORT')
    url += host
    if port:
        if str(port) == '443':
            url = 'https'+url
        elif str(port) == '80':
            url = 'http'+url
        else:
            # Assume we are running HTTP on a non-standard port
            url = 'http'+url+':%s' % port
    else:
        url = 'http'+url
    if script_name is None:
        url += urllib.quote(environ.get('SCRIPT_NAME',''))
    else:
        url += urllib.quote(script_name)
    if with_path_info:
        if path_info is None:
            url += urllib.quote(environ.get('PATH_INFO',''))
        else:
            url += urllib.quote(path_info)
    if with_query_string:
        if querystring is None:
            if environ.get('QUERY_STRING'):
                url += '?' + environ['QUERY_STRING']
        elif querystring:
            url += '?' + querystring
    return url

def load_form_config(
    app, 
    auth_conf, 
    app_conf=None,
    global_conf=None,
    prefix='authkit.method.form',
):
    app = RequireEnvironKey(
        app,
        'paste.auth_tkt.set_user',
        missing_error=(
            'Missing the key %(key)s from the environ. '
            'Have you added the cookie method after the form method?'
        )
    )
    template_conf = strip_base(auth_conf, 'template.')
    if template_conf:
        template_ = get_template(template_conf, prefix=prefix+'template.')
    else:
        template_ = template
    authenticate_conf = strip_base(auth_conf, 'authenticate.')
    app, authfunc, users = get_authenticate_function(
        app, 
        authenticate_conf, 
        prefix=prefix+'authenticate.', 
        format='basic'
    )
    charset = auth_conf.get('charset')
    method = auth_conf.get('method', 'post')
    action = auth_conf.get('action')
    user_data = auth_conf.get('userdata')
    if method.lower() not in ['get','post']:
        raise Exception('Form method should be GET or POST, not %s'%method)
    return app, {
        'authfunc': authfunc, 
        'template': template_, 
        'charset': charset, 
        'method': method,
        'action': action,
        'user_data': user_data or None,
    }, None

def make_form_handler(
    app, 
    auth_conf, 
    app_conf=None,
    global_conf=None,
    prefix='authkit.method.form', 
):
    app, auth_handler_params, user_setter_params = load_form_config(
        app, 
        auth_conf, 
        app_conf=None,
        global_conf=None,
        prefix='authkit.method.form',
    )
    app = MultiHandler(app)
    app.add_method(
        'form', 
        FormAuthHandler, 
        authfunc=auth_handler_params['authfunc'], 
        template=auth_handler_params['template'], 
        charset=auth_handler_params['charset'],
        method=auth_handler_params['method'],
        action=auth_handler_params['action'],
        user_data=auth_handler_params['user_data'],
    )
    app.add_checker('form', status_checker)
    return app

# Backwards compatbility
Form = FormAuthHandler

